src/lib/packetBuilder.ts

```diff
@@ -0,0 +1,68 @@
+import type { ProtocolVersion } from '../enums'
+import type { ISerializableCommand } from '../commands'
+
+export class PacketBuilder {
+	readonly #maxPacketSize: number
+	readonly #protocolVersion: ProtocolVersion
+
+	readonly #completedBuffers: Buffer[] = []
+
+	#finished = false
+	#currentPacketBuffer: Buffer
+	#currentPacketFilled: number
+
+	constructor(maxPacketSize: number, protocolVersion: ProtocolVersion) {
+		this.#maxPacketSize = maxPacketSize
+		this.#protocolVersion = protocolVersion
+
+		this.#currentPacketBuffer = Buffer.alloc(maxPacketSize)
+		this.#currentPacketFilled = 0
+	}
+
+	public addCommand(cmd: ISerializableCommand): void {
+		if (this.#finished) throw new Error('Packets have been finished')
+
+		if (typeof cmd.serialize !== 'function') {
+			throw new Error(`Command ${cmd.constructor.name} is not serializable`)
+		}
+
+		const rawName: string = (cmd.constructor as any).rawName
+		const payload = cmd.serialize(this.#protocolVersion)
+
+		const totalLength = payload.length + 8
+		if (totalLength > this.#maxPacketSize) {
+			throw new Error(`Comamnd ${cmd.constructor.name} is too large for a single packet`)
```

Comment:
Agreed, the default behaviour should probably be to put it into a new packet.

Would it be worth adding another argument into `sendCommands()` to allow for a consumer to get notified if the commands doesn't fit into one packet?
```typescript
export class AtemSocket {
// ...
/** 
 * sendCommands
 * @param commands commands to send
 * @param onlyOnePacket ensure commands are sent as a single packet. (throws if they doesn't fit)
 */
sendCommands(commands: Array<ISerializableCommand>, onlyOnePacket?: boolean)
```

Reply:
I'm not sure if a new parameter will be useful. These large commands will always be coming from the 'DataTransfer' system, so either way its behaviour will end up pretty much hardcoded one way.

But I think its a question of whether the transfer should immediately fail due to the packets being unsendable (I don't know how/if the logic will handle this anyway), or if they should be sent and left to timeout

Reply:
I think I prefer the idea of sending them anyway. I dont expect this to be a problem in practise (I dont know why anyone would need to change the maxPacketSize), so if it does then it can be dealt with if it does become a problem.